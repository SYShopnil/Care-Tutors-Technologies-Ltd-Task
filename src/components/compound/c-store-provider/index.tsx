"use client";
import { useRef } from "react";
import { Provider } from "react-redux";
import { todoStore, AppStore } from "../../../lib/store/store";

export default function CStoreProvider({
  children,
}: {
  children: React.ReactNode;
}) {
  const storeRef = useRef<AppStore>();
  if (!storeRef.current) {
    storeRef.current = todoStore();
  }
  return <Provider store={storeRef.current}>{children}</Provider>;
}
