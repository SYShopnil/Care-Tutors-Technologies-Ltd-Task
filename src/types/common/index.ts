export enum EDataTestId {
  SHeaderMain = "SHeaderMain",
  CToDoCreateUpdate = "CToDoCreateUpdate",
  CTodoItem = "CTodoItem",
  CInputForm = "CInputForm",
  CToDoList = "CToDoList",
  CButton = "CButton",
}
