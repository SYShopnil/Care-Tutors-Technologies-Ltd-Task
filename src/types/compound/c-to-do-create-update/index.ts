import { TodoItem } from "@src/lib/store/todoSlice";

export enum EToDoMode {
  CREATE = "CREATE",
  UPDATE = "UPDATE",
}

export interface ICEditTodo {
  todo?: Pick<TodoItem, "title" | "description" | "id">;
  onClose?: () => any;
  mode: EToDoMode;
}
