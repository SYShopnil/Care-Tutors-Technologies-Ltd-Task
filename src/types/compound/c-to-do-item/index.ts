import { TodoItem } from "@src/lib/store/todoSlice";

export interface ICTodoItem {
  todo: TodoItem;
}
