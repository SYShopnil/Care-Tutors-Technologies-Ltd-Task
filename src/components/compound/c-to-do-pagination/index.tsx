"use client";

import { IPagination } from "@src/types/compound/c-to-do-pagination";

export const CPaginationToDo = ({
  itemsPerPage,
  totalItems,
  paginate,
  currentPage,
}: IPagination) => {
  const pageNumbers = [];

  for (let i = 1; i <= Math.ceil(totalItems / itemsPerPage); i++) {
    pageNumbers.push(i);
  }

  return (
    <nav>
      <ul className="flex justify-center space-x-2">
        {pageNumbers.map((number) => (
          <li
            key={number}
            className={`px-4 py-2 border rounded ${
              currentPage === number ? "bg-[#7F4D4F] text-white" : "bg-white"
            }`}
          >
            <button onClick={() => paginate(number)}>{number}</button>
          </li>
        ))}
      </ul>
    </nav>
  );
};
