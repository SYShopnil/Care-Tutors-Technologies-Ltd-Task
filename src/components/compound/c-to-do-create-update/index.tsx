"use client";

import { CInputForm } from "@src/components/root/c-input-form";
import { addTodo, editTodo } from "@src/lib/store/todoSlice";
import {
  EToDoMode,
  ICEditTodo,
} from "@src/types/compound/c-to-do-create-update";
import { BtnColorSchema } from "@src/types/root";
import { useState } from "react";
import { useDispatch } from "react-redux";

export const CToDoCreateAndUpdate = ({ todo, onClose, mode }: ICEditTodo) => {
  const [title, setTitle] = useState(todo?.title || "");
  const [description, setDescription] = useState(todo?.description || "");
  const dispatch = useDispatch();
  const btnText = mode == EToDoMode.CREATE ? "Add ToDo" : "Update";
  const btnVariant =
    mode == EToDoMode.CREATE
      ? BtnColorSchema.SolidBgGrayTextViolet
      : BtnColorSchema.SolidBgWhiteTextGreen;
  const handleSubmit = (
    e: React.MouseEvent<HTMLElement> | React.FormEvent<HTMLFormElement>
  ) => {
    e.preventDefault();

    switch (mode) {
      case EToDoMode.CREATE: {
        if (title.trim() && description.trim()) {
          dispatch(addTodo({ title, description }));
          setTitle("");
          setDescription("");
        }
        break;
      }
      case EToDoMode.UPDATE: {
        todo && dispatch(editTodo({ id: todo.id, title, description }));
        onClose && onClose();
        break;
      }
    }
  };
  return (
    <CInputForm
      handleSubmit={handleSubmit}
      setDescription={setDescription}
      setTitle={setTitle}
      toDoData={{ description, title }}
      btnText={btnText}
      btnVariant={btnVariant}
    />
  );
};
