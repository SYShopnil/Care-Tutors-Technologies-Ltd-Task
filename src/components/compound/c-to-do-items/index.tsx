"use client";
import { useState } from "react";
import { useDispatch } from "react-redux";
import { CToDoCreateAndUpdate } from "../c-to-do-create-update";
import { TodoItem, deleteTodo } from "@src/lib/store/todoSlice";
import { EToDoMode } from "@src/types/compound/c-to-do-create-update";
import { Button } from "@src/components/root";
import { BtnColorSchema } from "@src/types/root";

interface ICTodoItem {
  todo: TodoItem;
}

export const CTodoItem = ({ todo }: ICTodoItem) => {
  const [isEditing, setIsEditing] = useState(false);
  const dispatch = useDispatch();

  const handleDelete = () => {
    dispatch(deleteTodo(todo.id));
  };

  return (
    <div className="p-5 bg-white shadow-lg rounded mb-4">
      {isEditing ? (
        <CToDoCreateAndUpdate
          todo={todo}
          onClose={() => setIsEditing(false)}
          mode={EToDoMode.UPDATE}
        />
      ) : (
        <div className="p-2">
          <h3 className="text-xl ">
            <span className="font-bold">Title: </span>
            {todo.title}
          </h3>
          <p>
            <span className="font-bold ">Description:</span> {todo.description}
          </p>
          <div className="flex space-x-4 mt-2">
            <Button
              btnText="Edit"
              colorSchema={BtnColorSchema.SolidBgGrayTextViolet}
              isArrow={true}
              clickHandler={() => setIsEditing(true)}
            />

            <Button
              btnText="Delete"
              colorSchema={BtnColorSchema.SolidBgWhiteTextGreen}
              isArrow={true}
              clickHandler={handleDelete}
            />
          </div>
        </div>
      )}
    </div>
  );
};
