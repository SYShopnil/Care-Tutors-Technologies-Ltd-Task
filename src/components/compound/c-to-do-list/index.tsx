"use client";
import { useState } from "react";
import { CPaginationToDo } from "../c-to-do-pagination";
import { CTodoItem } from "../c-to-do-items";
import { TodoItem } from "@src/lib/store/todoSlice";
import { useAppSelector } from "@src/lib/store/hook";
import { showItemLimitPerPage } from "@src/config";
import { SLoading } from "@src/components/root";

const CTodoList = () => {
  const todos = useAppSelector((state) => state.todos);
  const [currentPage, setCurrentPage] = useState(1);

  const indexOfLastItem = currentPage * showItemLimitPerPage;
  const indexOfFirstItem = indexOfLastItem - showItemLimitPerPage;
  const currentTodos: TodoItem[] = todos.slice(
    indexOfFirstItem,
    indexOfLastItem
  );

  const paginate: (pageNumber: number) => void = (pageNumber) =>
    setCurrentPage(pageNumber);

  return (
    <div>
      {currentTodos.length ? (
        <>
          {currentTodos.map((todo) => (
            <CTodoItem key={todo.id} todo={todo} />
          ))}
          <CPaginationToDo
            itemsPerPage={showItemLimitPerPage}
            totalItems={todos.length}
            paginate={paginate}
            currentPage={currentPage}
          />
        </>
      ) : (
        <div>
          <SLoading text="No ToDo Found!!" />
        </div>
      )}
    </div>
  );
};
export default CTodoList;
