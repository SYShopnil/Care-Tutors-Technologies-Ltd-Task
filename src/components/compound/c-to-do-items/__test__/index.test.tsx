import { render, screen } from "@testing-library/react";
import { EDataTestId } from "@src/types/common";
import CStoreProvider from "../../c-store-provider";
import { CTodoItem } from "..";
import { ICTodoItem } from "@src/types/compound/c-to-do-item";

const mock_to_do_item: ICTodoItem = {
  todo: {
    completed: true,
    description: "Hello this is completed",
    id: "2123",
    title: "This is demo title",
  },
};
describe("When To Do Item Rendered", () => {
  it("Expect render successfully", async () => {
    render(
      <CStoreProvider>
        <CTodoItem {...mock_to_do_item} />
      </CStoreProvider>
    );
    const myElement = screen.getByRole(EDataTestId.CTodoItem);
    expect(myElement).toMatchSnapshot();
  });
});
