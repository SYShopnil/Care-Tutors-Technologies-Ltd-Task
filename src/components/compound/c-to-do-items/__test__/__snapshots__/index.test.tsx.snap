// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When To Do Item Rendered Expect render successfully 1`] = `
<div
  class="p-5 bg-white shadow-lg rounded mb-4"
  role="CTodoItem"
>
  <div
    class="p-2"
  >
    <h3
      class="text-xl "
    >
      <span
        class="font-bold"
      >
        Title: 
      </span>
      This is demo title
    </h3>
    <p>
      <span
        class="font-bold "
      >
        Description:
      </span>
       
      Hello this is completed
    </p>
    <div
      class="flex space-x-4 mt-2"
    >
      <span
        class="cursor-pointer"
        role="CButton"
      >
        <span
          class="flex justify-center items-center  "
        >
          <span
            class="text-base md:text-base_2  flex justify-center items-center px-[1rem] py-[0.5rem] rounded-[0.563rem] w-[100%] h-[100%] bg-[#EDEDED] hover:bg-[#7A67EE]  duration-[0.5s] text-[#7A67EE] hover:text-[#EDEDED]   border-[0.063rem] hover:border-[#EDEDED] border-[#7A67EE]   "
          >
            <span>
              Edit
            </span>
            <span
              class="ml-[1rem] h-[1.5rem] w-[1.5rem]  flex justify-center items-center"
            >
              <svg
                fill="#7A67EE"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 8c137 0 248 111 248 248S393 504 256 504 8 393 8 256 119 8 256 8zM140 300h116v70.9c0 10.7 13 16.1 20.5 8.5l114.3-114.9c4.7-4.7 4.7-12.2 0-16.9l-114.3-115c-7.6-7.6-20.5-2.2-20.5 8.5V212H140c-6.6 0-12 5.4-12 12v64c0 6.6 5.4 12 12 12z"
                />
              </svg>
            </span>
          </span>
        </span>
      </span>
      <span
        class="cursor-pointer"
        role="CButton"
      >
        <span
          class="flex justify-center items-center  "
        >
          <span
            class="text-base md:text-base_2  flex justify-center items-center px-[1rem] py-[0.5rem] rounded-[0.563rem] w-[100%] h-[100%] bg-[#EBF5EC] hover:bg-[#7F4D4F]  duration-[0.5s] text-[#7F4D4F] hover:text-[#EBF5EC]   border-[0.063rem] hover:border-[#EBF5EC] border-[#7F4D4F]   "
          >
            <span>
              Delete
            </span>
            <span
              class="ml-[1rem] h-[1.5rem] w-[1.5rem]  flex justify-center items-center"
            >
              <svg
                fill="#7F4D4F"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 8c137 0 248 111 248 248S393 504 256 504 8 393 8 256 119 8 256 8zM140 300h116v70.9c0 10.7 13 16.1 20.5 8.5l114.3-114.9c4.7-4.7 4.7-12.2 0-16.9l-114.3-115c-7.6-7.6-20.5-2.2-20.5 8.5V212H140c-6.6 0-12 5.4-12 12v64c0 6.6 5.4 12 12 12z"
                />
              </svg>
            </span>
          </span>
        </span>
      </span>
    </div>
  </div>
</div>
`;
